### PR TITLE
ogent: support enums

### LINKED ISSUE #51: Ogent doesn't work with ENUM types.
When I try to use Ogent with Enum , I am getting Errors.
I have hosted an example at https://github.com/chasak/enumeg
Here is part of schema file.
```
func (Video) Fields() []ent.Field {
	return []ent.Field{
		field.String("title"),
		field.Enum("videotype").Values("live", "video", "playlist"),
	}
}
```
This generates the following response
```
func NewGroupVideosList(e *ent.Video) *GroupVideosList {
	if e == nil {
		return nil
	}
	return &GroupVideosList{
		ID:          e.ID,
		Title:       e.Title,
		Videotype:   e.Videotype,
	}
}
```
The error is on `Videotype:   e.Videotype,` line due to the mismatched types. Take a look at `GroupVideosList`

```
type GroupVideosList struct {
	ID          int                      "json:\"id\""
	Title       string                   "json:\"title\""
	Videotype   GroupVideosListVideotype "json:\"videotype\""
}

type GroupVideosListVideotype string

const (
	GroupVideosListVideotypeLive     GroupVideosListVideotype = "live"
	GroupVideosListVideotypeVideo    GroupVideosListVideotype = "video"
	GroupVideosListVideotypePlaylist GroupVideosListVideotype = "playlist"
)

```

and `ent.Video`

```
type Video struct {
	config `json:"-"`
	// ID of the ent.
	ID int `json:"id,omitempty"`
	// Title holds the value of the "title" field.
	Title string `json:"title,omitempty"`
	// Videotype holds the value of the "videotype" field.
	Videotype video.Videotype `json:"videotype,omitempty"`
	// Edges holds the relations/edges for other nodes in the graph.
	// The values are being populated by the VideoQuery when eager-loading is set.
	Edges        VideoEdges `json:"edges"`
	group_videos *int
}

// Videotype defines the type for the "videotype" enum field.
type Videotype string

// Videotype values.
const (
	VideotypeLive     Videotype = "live"
	VideotypeVideo    Videotype = "video"
	VideotypePlaylist Videotype = "playlist"
)

```
The error is happeing due to not doing explicit typecasting and can be solved by 
```Videotype:   GroupVideosListVideotype(e.Videotype),```

```
func NewGroupVideosList(e *ent.Video) *GroupVideosList {
	if e == nil {
		return nil
	}
	return &GroupVideosList{
		ID:          e.ID,
		Title:       e.Title,
		Videotype:   GroupVideosListVideotype(e.Videotype),
	}
}
```

Same Issue with handler Function

```
func (h *OgentHandler) CreateVideo(ctx context.Context, req CreateVideoReq) (CreateVideoRes, error) {
	b := h.client.Video.Create()
	// Add all fields.
	b.SetTitle(req.Title)
	b.SetVideotype(req.Videotype)
}
```
also here

```
func (h *OgentHandler) UpdateVideo(ctx context.Context, req UpdateVideoReq, params UpdateVideoParams) (UpdateVideoRes, error) {
	b := h.client.Video.UpdateOneID(params.ID)
	// Add all fields.
	if v, ok := req.Title.Get(); ok {
		b.SetTitle(v)
	}
	if v, ok := req.Videotype.Get(); ok {
		b.SetVideotype(v)
	}

```
Thanks.

### PR DESCRIPTION
**Pull request by masseelch**
Add support for enums. 

Fix #51.

### CODE REVIEW

File: template.go [hunk 1]
```diff
@@ -176,16 +177,27 @@ const (
 )
 
 // setFieldExpr returns a Go expression to set the field on a response.
-func setFieldExpr(f *gen.Field, ident string) (string, error) {
+func setFieldExpr(f *gen.Field, schema, ident string) (string, error) {
 	if !f.Optional {
-		return fmt.Sprintf("%s: %s.%s", f.StructField(), ident, f.StructField()), nil
+		expr := fmt.Sprintf("%s.%s", ident, f.StructField())
+		if f.IsEnum() {
+			expr = convertTo(schema+f.StructField(), expr)
```
Comment (a8m) [inline template.go:184]: I think it won't work if you have a custom `GoType`, and need to use `gen.Field.Type.String`.

File: template/create.tmpl [hunk 2]
```diff
@@ -4,10 +4,14 @@
 	{{- range $f := $.Fields }}
 		{{- if $f.Optional }}
 			if v, ok := req.{{ $f.StructField}}.Get(); ok {
-				b.Set{{ $f.StructField }}(v)
+				{{ $x := "v" }}
+				{{ if $f.IsEnum }}{{ $x = convertTo $f.Type.String $x }}{{ end }}
```
Comment (a8m) [inline template/create.tmpl:8]:
```suggestion
				{{- $x := "v" }}
				{{- if $f.IsEnum }}{{ $x = convertTo $f.Type.String $x }}{{ end }}
```

File: template/create.tmpl [hunk 3]
```diff
@@ -4,10 +4,14 @@
...
+				b.Set{{ $f.StructField }}({{ $x }})
 			}
 		{{- else }}
-			b.Set{{ $f.StructField }}(req.{{ $f.StructField }})
+			{{ $x := print "req." $f.StructField }}
+			{{ if $f.IsEnum }}{{ $x = convertTo $f.Type.String $x }}{{ end }}
```
Comment (a8m) [inline template/create.tmpl:13]: ditto

File: template/update.tmpl [hunk 4]
```diff
@@ -6,7 +6,9 @@
 	{{- range $f := $.Fields }}
 		{{- if not $f.Immutable }}
 			if v, ok := req.{{ $f.StructField}}.Get(); ok {
-				b.Set{{ $f.StructField }}(v)
+				{{ $x := "v" }}
+				{{ if $f.IsEnum }}{{ $x = convertTo $f.Type.String $x }}{{ end }}
```
Comment (a8m) [inline template/update.tmpl:10]: ditto

File: internal/integration/ogent/ent/ogent/ogent.go [hunk 5]
```diff
@@ -577,11 +593,21 @@ func (h *OgentHandler) UpdateUser(ctx context.Context, req UpdateUserReq, params
...
 		b.SetAge(v)
 	}
+	if v, ok := req.Sex.Get(); ok {
+
+		b.SetSex(user.Sex(v))
+	}
+	if v, ok := req.Gender.Get(); ok {
+
+		b.SetGender(user.Gender(v))
+	}
```
Comment (a8m) [inline internal/integration/ogent/ent/ogent/ogent.go:610]: See my comments to avoid all these spaces.

File: template.go (hunk 1 above)
Comment (masseelch) [reply template.go:184]: It actually does work with `GoType` on enums. But `GoType`s don't work on other stuff. But that is to be fixed in another PR.

**Review by a8m — APPROVED** [review]

**Review by tdakkota — APPROVED** [review]

File: template/response.tmpl [hunk 6]
```diff
@@ -14,7 +14,7 @@ import "{{ $.Config.Package }}"
         }
         return &{{ $n }}{
             {{- range $f := $v.Fields }}
-                {{ setFieldExpr $f "e" }},
+                {{ $f.StructField}}: {{ setFieldExpr $f $n "e" }},
```
Comment (chaychoong) [inline template/response.tmpl:17]: I'm using enums using a custom GoType, and I am getting this error
```
running ent codegen: execute template "ogent/responses": template: response.tmpl:17:40: executing "ogent/responses" at <setFieldExpr $f $n "e">: error calling setFieldExpr: unexpected type: ""
```

File: template.go (hunk 1 above)
Comment (masseelch) [reply template.go:184]: Hey @chaychoong, can you create an issue and provide me an example of the type you are using?

Comment (masseelch) [reply template.go:184]: nvm, I found it. Will have a fix out soon.

**Pull request merged**